File: QID-1226-MVAportfol/MVAportfol.r

# clear all variables
rm(list = ls(all = TRUE))
graphics.off()

# load data
ibm   = read.csv("ibm.csv")
apple = read.csv("apple.csv")
bac   = read.csv("bac.csv")
ford  = read.csv("ford.csv")
ed    = read.csv("ed.csv")
ms    = read.csv("ms.csv")

# compute returns for IBM
y1 = ibm[, 5]
a  = 0
i  = 1
while (i <= 120) {
    i    = i + 1
    a[i] = (y1[i] - y1[i - 1])/y1[i]
}
x1 = a[2:121] 	

# compute returns for Apple
y2 = apple[, 5]
b  = 0
i  = 1
while (i <= 120) {
    i    = i + 1
    b[i] = (y2[i] - y2[i - 1])/y2[i]
}
x2 = b[2:121] 	

# compute returns for Bank of America Corporation
y3 = bac[, 5]
d  = 0
i  = 1
while (i <= 120) {
    i    = i + 1
    d[i] = (y3[i] - y3[i - 1])/y3[i]
}
x3 = d[2:121]

# compute returns for Forward Industries
y4 = ford[, 5]
f  = 0
i  = 1
while (i <= 120) {
    i    = i + 1
    f[i] = (y4[i] - y4[i - 1])/y4[i]
}
x4 = f[2:121]

# compute returns for Consolidated Edison
y5 = ed[, 5]
g  = 0
i  = 1
while (i <= 120) {
    i    = i + 1
    g[i] = (y5[i] - y5[i - 1])/y5[i]
}
x5 = g[2:121]

# compute returns for Morgan Stanley
y6 = ms[, 5]
h  = 0
i  = 1
while (i <= 120) {
    i    = i + 1
    h[i] = (y6[i] - y6[i - 1])/y6[i]
}
x6 = h[2:121]

ix  = c(1:6)                                # choose assets to use
x   = cbind(x1, x2, x3, x4, x5, x6)         # MVAportfolio of all assets
s1  = solve(cov(x))                         # inverse of empirical variance
one = rep(1, length(ix))                    # vector of ones
c2  = (s1 %*% one)/rep(t(one) %*% s1 %*% one, length(s1 %*% one))    # c2 weight
c1  = one/rep(sum(one), length(one))        # c1 weight
q1  = x %*% c1                              # Optimal MVAportfol returns 
q2  = x %*% c2                              # Nonoptimal MVAportfol returns 

t  = c(1:120)
d1 = cbind(t, q1)
d2 = cbind(t, q2)

# plot
par(mfrow = c(2, 1))
par(oma = c(0, 0, 0, 8))
plot(d1, type = "l", col = "blue", ylab = "Y", xlab = "X", main = "Equally Weighted Portfolio", 
    ylim = c(-0.6, 0.6), cex.lab = 1.4, cex.axis = 1.4, cex.main = 1.4)
mtext("Weights", side = 4, line = 4, at = 1.2, las = 2, font = 1)
mtext(toString(c(sprintf("%.3f", c1[1]), "IBM")), side = 4, line = 4, at = 0.6, las = 2)
mtext(toString(c(sprintf("%.3f", c1[2]), "Apple")), side = 4, line = 4, at = 0.45, 
    las = 2)
mtext(toString(c(sprintf("%.3f", c1[3]), "BAC")), side = 4, line = 4, at = 0.3, las = 2)
mtext(toString(c(sprintf("%.3f", c1[4]), "Ford")), side = 4, line = 4, at = 0.15, 
    las = 2)
mtext(toString(c(sprintf("%.3f", c1[5]), "Edison")), side = 4, line = 4, at = 0, 
    las = 2)
mtext(toString(c(sprintf("%.3f", c1[6]), "Stanley")), side = 4, line = 4, at = -0.15, 
    las = 2)
plot(d2, type = "l", col = "blue", ylab = "Y", xlab = "X", main = "Optimal Weighted Portfolio", 
    ylim = c(-0.6, 0.6), cex.lab = 1.4, cex.axis = 1.4, cex.main = 1.4)
mtext("Weights", side = 4, line = 4, at = 1.2, las = 2, font = 1)
mtext(toString(c(sprintf("%.3f", c2[1]), "IBM")), side = 4, line = 4, at = 0.6, las = 2)
mtext(toString(c(sprintf("%.3f", c2[2]), "Apple")), side = 4, line = 4, at = 0.45, 
    las = 2)
mtext(toString(c(sprintf("%.3f", c2[3]), "BAC")), side = 4, line = 4, at = 0.3, las = 2)
mtext(toString(c(sprintf("%.3f", c2[4]), "Ford")), side = 4, line = 4, at = 0.15, 
    las = 2)
mtext(toString(c(sprintf("%.3f", c2[5]), "Edison")), side = 4, line = 4, at = 0, 
    las = 2)
mtext(toString(c(sprintf("%.3f", c2[6]), "Stanley")), side = 4, line = 4, at = -0.15, 
    las = 2) 

File: QID-1226-MVAportfol/MVAportfol.py
# Works on pandas 2.0.0, numpy 1.24.2, matplotlib 3.7.1
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Load data
ibm = pd.read_csv("ibm.csv")
apple = pd.read_csv("apple.csv")
bac = pd.read_csv("bac.csv")
ford = pd.read_csv("ford.csv")
ed = pd.read_csv("ed.csv")
ms = pd.read_csv("ms.csv")

# Function to compute returns
def compute_returns(df):
    y = df.iloc[:, 4]
    a = [0] 
    for i in range(1, 121):  
        a.append((y[i] - y[i - 1]) / y[i])
    return a[1:]

# Compute returns for each company
x1 = compute_returns(ibm)
x2 = compute_returns(apple)
x3 = compute_returns(bac)
x4 = compute_returns(ford)
x5 = compute_returns(ed)
x6 = compute_returns(ms)

# Combine all assets
x = np.column_stack((x1, x2, x3, x4, x5, x6))

# Calculate portfolio weights
s1 = np.linalg.inv(np.cov(x.T))  # Inverse of empirical variance
one = np.ones(6)  # Vector of ones
c2 = np.dot(s1, one) / np.dot(np.dot(one.T, s1), one)  # Optimal weights
c1 = one / one.sum()  # Equally weighted

# Calculate portfolio returns
q1 = np.dot(x, c1)  # Equally weighted portfolio returns
q2 = np.dot(x, c2)  # Optimal weighted portfolio returns

# Time Series
t = np.arange(1, 121)
ts1 = pd.DataFrame({'Time': t, 'Returns': q1})
ts2 = pd.DataFrame({'Time': t, 'Returns': q2})

# Plotting
plt.figure(figsize=(8, 8))

stock_names = ['IBM', 'Apple', 'BAC', 'Ford', 'Edison', 'Stanley']

# Equally Weighted Portfolio
plt.subplot(2, 1, 1)
plt.plot(ts1['Time'], ts1['Returns'], color='blue', linewidth=0.7)
plt.title('Equally Weighted Portfolio', fontweight='bold')
plt.xlabel('X')
plt.ylabel('Y')
plt.ylim(-0.65, 0.65)
plt.yticks([-0.6, -0.4, -0.2, 0, 0.2, 0.4, 0.6], ['-0.6', '', '', '0', '', '', '0.6'])
weights_text = "Weights\n\n\n\n" + "\n".join([f"{weight:.3f}, {name}" for weight, name in zip(c1, stock_names)])
plt.text(ts1['Time'].iloc[-1] + 15, 0, weights_text, verticalalignment='center')

# Optimal Weighted Portfolio
plt.subplot(2, 1, 2)
plt.plot(ts2['Time'], ts2['Returns'], color='blue', linewidth=0.7)
plt.title('Optimal Weighted Portfolio', fontweight='bold')
plt.xlabel('X')
plt.ylabel('Y')
plt.ylim(-0.65, 0.65)
plt.yticks([-0.6, -0.4, -0.2, 0, 0.2, 0.4, 0.6], ['-0.6', '', '', '0', '', '', '0.6'])
weights_text = "Weights\n\n\n\n" + "\n".join([f"{weight:.3f}, {name}" for weight, name in zip(c2, stock_names)])
plt.text(ts2['Time'].iloc[-1] + 15, 0, weights_text, verticalalignment='center')

plt.tight_layout()
plt.show()